include/exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.cpp

```diff
@@ -404,7 +404,10 @@ SpatialOperatorBase<dim, Number>::initialize_operators(std::string const & dof_i
   viscous_kernel_data.viscosity_is_variable        = param.use_turbulence_model;
   viscous_kernel_data.variable_normal_vector       = param.neumann_with_variable_normal_vector;
   viscous_kernel = std::make_shared<Operators::ViscousKernel<dim, Number>>();
-  viscous_kernel->reinit(*matrix_free, viscous_kernel_data, get_dof_index_velocity());
+  viscous_kernel->reinit(*matrix_free,
+                         viscous_kernel_data,
+                         get_dof_index_velocity(),
+                         get_quad_index_velocity_linearized());
```

Comment:
I think this line is problematic. `ViscousKernel` is used for both the `ViscousOperator` (used e.g. to evaluate the nonlinear residual of the incompressible Navier-Stokes equations, using the standard quadrature rule with k+1 points for the viscous term) and the `MomentumOperator` (the (1,1)-block in the linearized system of equations). For the momentum operator (including the convective term), all operators (mass, convection, diffusion) have to use the same quadrature rule, so one can choose different quadrature rules (3/2-overintegration vs. standard quadrature rule with k+1 points). Hence, for the viscous kernel it is in general not clear which quadrature rule to use. 

- In the most general case, viscous kernel would have to support different quadrature rules and `VariableCoefficients` data for these different quadrature rules (if we want to support all the present functionality and parameter choices). This would increase storage requirements, but could be realized easily from an implementational point of view.
-  Another solution could be that the viscous operator always has to use the same quadrature rule as the convective operator in case a fully-implicit, monolithic solution approach is selected to solve the incompressible Navier-Stokes problem (increasing computational costs). This solution would only make sense if we no longer allow to select different quadrature rules for the convective term in the nonlinear residual and in the linearized operator (which is possible in the current code version).
-  Yet another solution could be to remove again the merging of kernels in `MomentumOperator`, which we introduced some time ago to improve performance compared to a sequential matrix-free evaluation with separate loops for mass, convective, viscous terms). Performance would deteriorate in this case. The benefit would be a strictly modular design with independent operators and a unique quadrature rule for the viscous term. Note, however, that such changes would require to rethink the multigrid preconditioner implementation.

I think the best / easiest solution is currently to use `get_quad_index_velocity_linear()` here and to assert that the standard quadrature is used for the linearized momentum operator if `viscous_kernel_data.viscosity_is_variable == true` with a comment that one would otherwise have to initialize `VariableCoefficients` for an additional quadrature rule using over-integration.

Overall, this points to a general design problem in ExaDG. Which class is responsible for quad-indices and the handling of variable coefficients? `Kernel` vs. `Operator`? @btemuer if you are interested, we could discuss in more detail what a good solution to this design problem could be. @richardschu this could also be relevant for your applications with variable viscosity. @kronbichler what do you think? You have accompanied the developments from the very beginning.

Reply:
Yep, finding this as the cause of a related error message when applying the `MomentumOperator` in the nonlinear case  took me more time than it should have ... in a PR I will open today, I put some assertions for the case of variable viscosity, since the default is using `QuadratureRuleLinearization::Overintegration32k`, leading to an index access out of range when evaluating the momentum operator (same as you propose; fits for an intermediate solution until we support one of the options you listed).

Regarding your posed concepts, I would go for option 1 and accept the increased storage efforts in case one wants to solve for the velocity in a nonlinear fashion. I am not sure if I would sacrifice better performance for the possibility of overintegration? The question is, how important overintegration really is? In this regard, I have unfortunately little to no experience, so what is your take on that? I have run my tests with `QuadratureRuleLinearization::Standard` so far and it seemed fine ... but again, I did not do intense testing. 

Comparing Options 2 and 3, how does the increased effort of using 3/2k instead of Standard compare to the merging or individual application of the kernels in the `MomentumOperator`?

Reply:
To start with, let me note that **over-integration** is necessary in cases the simulation is under-resolved, because then we might aliasing errors that can make the simulation break. Thus, there is a need to support over-integration in general, but there might be applications where we can get around with fewer points (possible in the current setting). Hence, one cannot really compare the effort of 3k/2 vs standard and merging/not merging because the former is an accuracy requirement, whereas the latter is a software/performance issue without influence on the visible accuracy.

Regarding the design, my position is that option 3 of having separate global operators for different terms in the Navier-Stokes operator as the only way to implement operators in ExaDG would contradict the ambition to be an HPC code on CPU-type hardware. On GPUs, I could see the split making sense in certain implementations (in case one wants to save registers and would need to go out for global memory to get the data again anyway), but trends for GPUs also go towards caches. But who knows about the future, and performance modeling and measurement will be an essential ingredient on that path to choose whatever fits the hardware and available software best.
This does not mean that having the possibility to build operators from building blocks is wrong, it's quite handy for development, e.g. parts that are rarely used and thus not performance critical. But for important operators that do this in inner loops, I would be willing to spend considerable resources to find, implement and maintain better data locality variants.
Along this background, I think the first variant is the most appropriate one of the three options at this stage. It is not pretty and this is certainly requiring more improvement in the future. As a general direction, I suggest we prepare ourselves for being able to use standard quadrature formulas also for convective terms in "inner loops" (e.g. preconditioners), but over-integration and similar things in more outer loops. This fits very well in the first concept expressed by @nfehn above, as the memory consumption is likely not our primary concern. The overall goal is to perform the bulk of the work with approximate (but cheap) operator evaluation, say standard quadrature, and use the actually desired operator with best accuracy only rarely. This will require additional research and additional changes in the software layers once we go for it, but I believe that there are substantial gains to be made this way. This is one ingredient on our path to the future, we should consider also other options, including skew-symmetric formulations of convective terms or the like, to identify the best compromise in terms of accuracy & stability versus run-time. In particular with implicit time stepping, we have plenty of options.

Reply:
One should add to the discussion that we have observed in the past that using a standard quadrature rule for the linearized momentum operator reduces the costs for the operator evaluation, but increases iteration counts noticably. From these numerical experiments, it was not clear whether standard quadrature pays off in terms of overall efficiency. However, we left this option in the code so that one can later again easily investigate this topic.

Reply:
Option 1 seems to be preferred, so let's realize this one. Since there has been a bug in the code before the present PR, we should first fix this bug by adding an assert to the code (as described above), so that it is clear for the present PR which quadrature rule to plug in for the viscous kernel.

Afterwards, in a follow-up PR, we can realize the viscous kernel operating on multiple quadrature rules.

Reply:
Alright, I'm on it👍

Reply:
> Yep, finding this as the cause of a related error message when applying the MomentumOperator in the nonlinear case took me more time than it should have ... 

This should be reported in form of an issue as soon as such problems are encountered.

> in a PR I will open today, I put some assertions for the case of variable viscosity, since the default is using

Bugfixes should be proposed as a separate PR and not as a minor change within a long PR with several thousand lines of code being changed.

To make progress with the present PR, I suggest that @btemuer adds an assert where the MomentumOperator is set up, or @richardschu adds the assert in a separate PR, so that @btemuer can select the proper quadrature rule for the ViscousKernel to proceed with the present PR. 

All other changes such as multiple quadrature rules, nonlinear viscosity models etc. are to be discussed in the future.

Reply:
I've already added an assert in the parameter check thinking it would be equivalent. I also selected the proper quadrature rule. Can you check the latest commit?

Reply:
looks good